File: src/lib/domain/reader.ts
export type DiscoveryReaderCandidate = {
  id: string;
  sourceProvider?: string;
  sourceId?: string;
  title: string;
  authors: string[];
  venue: string;
  year: number;
  citations: number;
  tags: string[];
  abstract?: string;
  externalUrl?: string;
  pdfUrl?: string;
  doi?: string;
  arxivId?: string;
};

export type ReaderExtractor = "docling" | "mineru" | "marker" | "grobid" | "pdfium_basic";

export type ReaderPage = {
  pageIndex: number;
  width: number;
  height: number;
};

export type ReaderBlock = {
  id: string;
  pageIndex: number;
  blockIndex: number;
  readingOrder: number;
  kind: "title" | "authors" | "heading" | "paragraph" | "caption" | "figure" | "table" | "equation";
  text?: string;
  assetId?: string;
  sourceStart?: number;
  sourceEnd?: number;
  bbox?: [number, number, number, number];
};

export type ReaderSpan = {
  id: string;
  blockId: string;
  pageIndex: number;
  text: string;
  sourceStart: number;
  sourceEnd: number;
  bbox: [number, number, number, number];
};

export type ReaderAsset = {
  id: string;
  paperId: string;
  sourceId: string;
  extractionId: string;
  kind: "page_image" | "embedded_image" | "figure" | "table" | "equation";
  pageIndex: number;
  bbox?: [number, number, number, number];
  localPath: string;
  caption?: string;
};

export type ExtractedDocument = {
  paperId: string;
  sourceId: string;
  extractionId: string;
  annotationSourceId: string;
  extractor: ReaderExtractor;
  extractorVersion: string;
  sourceText: string;
  pages: ReaderPage[];
  blocks: ReaderBlock[];
  spans: ReaderSpan[];
  assets: ReaderAsset[];
};

export type ReaderParagraph = {
  id: string;
  kind: "heading" | "paragraph";
  text: string;
  highlight?: "soft" | "strong";
};

export type ReaderTextBlock = {
  id: string;
  kind: "title" | "authors" | "heading" | "paragraph";
  text: string;
  sourceStart: number;
  highlight?: "soft" | "strong";
};

export type ReaderTextSelection = {
  sourceId: string;
  startOffset: number;
  endOffset: number;
  selectedText: string;
  anchorKind?: "text_offset" | "pdf_rect";
  pageIndex?: number;
  rectsJson?: string;
  quoteContext?: string;
};

export type PdfRect = {
  x: number;
  y: number;
  width: number;
  height: number;
};

export type ReaderMark = {
  id: string;
  paragraphId: string;
  kind: "note" | "question" | "highlight";
  body: string;
  createdLabel: string;
};

export type ReaderDocument = {
  paperId: string;
  sourceId: string;
  extractionId?: string;
  annotationSourceId?: string;
  sourceText: string;
  title: string;
  authors: string[];
  venue: string;
  year: number;
  identifier: string;
  citationKey: string;
  tags: string[];
  pdfLocalPath?: string;
  pdfSourceUrl?: string;
  pdfError?: string;
  /** "acquiring" while a background download is in flight (RFC 0051). */
  pdfStatus?: string;
  pages: ReaderPage[];
  blocks: ReaderBlock[];
  spans: ReaderSpan[];
  assets: ReaderAsset[];
  textBlocks: ReaderTextBlock[];
  paragraphs: ReaderParagraph[];
  marks: ReaderMark[];
};

File: src-tauri/src/domain/reader.rs
use serde::{Deserialize, Serialize};

/// Discovery candidate payload sent to the Reader before a paper is saved.
///
/// This keeps the Reader command boundary small: the frontend can pass the
/// metadata it already has from Discover, and the backend can adapt it into the
/// normal `ReaderDocument` shape without creating a durable library row first.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DiscoveryReaderCandidate {
    pub id: String,
    pub source_provider: Option<String>,
    pub source_id: Option<String>,
    pub title: String,
    pub authors: Vec<String>,
    pub venue: String,
    pub year: i32,
    pub citations: i32,
    pub tags: Vec<String>,
    #[serde(rename = "abstract")]
    pub abstract_text: Option<String>,
    pub external_url: Option<String>,
    pub pdf_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderTextBlock {
    pub id: String,
    pub kind: String,
    pub text: String,
    pub source_start: i64,
    pub highlight: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderParagraph {
    pub id: String,
    pub kind: String,
    pub text: String,
    pub highlight: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderMark {
    pub id: String,
    pub paragraph_id: String,
    pub kind: String,
    pub body: String,
    pub created_label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderPage {
    pub page_index: i32,
    pub width: f64,
    pub height: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderBlock {
    pub id: String,
    pub page_index: i32,
    pub block_index: i32,
    pub reading_order: i32,
    pub kind: String,
    pub text: Option<String>,
    pub asset_id: Option<String>,
    pub source_start: Option<i64>,
    pub source_end: Option<i64>,
    pub bbox_json: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderSpan {
    pub id: String,
    pub block_id: String,
    pub page_index: i32,
    pub text: String,
    pub source_start: i64,
    pub source_end: i64,
    pub bbox_json: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderAsset {
    pub id: String,
    pub paper_id: String,
    pub source_id: String,
    pub extraction_id: String,
    #[serde(rename = "kind")]
    pub asset_kind: String,
    pub page_index: i32,
    pub bbox_json: Option<String>,
    pub local_path: String,
    pub caption: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderDocument {
    pub paper_id: String,
    pub source_id: String,
    pub extraction_id: Option<String>,
    pub annotation_source_id: Option<String>,
    pub title: String,
    pub authors: Vec<String>,
    pub venue: String,
    pub year: i32,
    pub identifier: String,
    pub citation_key: String,
    pub tags: Vec<String>,
    pub pdf_local_path: Option<String>,
    pub pdf_source_url: Option<String>,
    pub pdf_error: Option<String>,
    pub source_text: String,
    pub pages: Vec<ReaderPage>,
    pub blocks: Vec<ReaderBlock>,
    pub spans: Vec<ReaderSpan>,
    pub assets: Vec<ReaderAsset>,
    pub text_blocks: Vec<ReaderTextBlock>,
    pub paragraphs: Vec<ReaderParagraph>,
    pub marks: Vec<ReaderMark>,
}
